source/simulator/assembly.cc

```diff
@@ -1816,6 +2808,31 @@ namespace aspect
                                                     data.local_dof_indices,
                                                     system_matrix,
                                                     system_rhs);
+
+    if (advection_field.is_discontinuous(introspection))
+      {
+        const unsigned int test_output = 0;
+        for (unsigned int f=0; f<GeometryInfo<dim>::max_children_per_face
+             * GeometryInfo<dim>::faces_per_cell; ++f)
```

Comment:
This loop is doing a lot of copying: In 3d, this is 24 (sub-)faces, and each matrix has size at least 116 (if no compositional fields). But in most cases, only a subset of these 24 (sub-)face matrices will have been filled. Can we somehow keep track of which ones have been filled, and only copy those? For example, by keeping a flag in the copydata structure for each matrix, that is reset to false at the beginning of the loop over the faces, and set to true everytime you write something into it?

The entries you copy are also mostly zero (namely everything in the local matrices that has to do with velocities and pressures), and we could probably avoid some copying. But that's a pre-existing issue, and so maybe not important.